huaweicloud/services/servicestage/resource_huaweicloud_servicestage_application.go

```diff
@@ -0,0 +1,284 @@
+			return err
+		}
+	}
+	return nil
+}
+
+func resourceApplicationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.ServiceStageV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating ServiceStage v2 client: %s", err)
+	}
+
+	opt := buildApplicationCreateOpts(d, config.GetEnterpriseProjectID(d))
+	log.Printf("[DEBUG] The CreateOpt of ServiceStage application is: %v", opt)
+	resp, err := applications.Create(client, opt)
+	if err != nil {
+		return diag.Errorf("error creating ServiceStage application: %s", err)
+	}
+
+	d.SetId(resp.ID)
+
+	err = createEnvironmentVariables(client, d.Id(), d.Get("environment").(*schema.Set))
+	if err != nil {
+		return diag.Errorf("error creating environment variable: %s", err)
+	}
+
+	return resourceApplicationRead(ctx, d, meta)
+}
+
+func flattenEnvironmentVariables(d *schema.ResourceData, configs []applications.ConfigResp) []map[string]interface{} {
+	// Store all variables of the same environment in the corresponding list, and generate a map with the environment ID
+	// as the key name.
+	result := make([]map[string]interface{}, 0)
+	for _, v := range configs {
+		if v.ApplicationId != d.Id() || len(v.Configuration.EnvVariables) < 1 {
+			continue
+		}
+
+		variables := make([]map[string]interface{}, len(v.Configuration.EnvVariables))
+		for i, variable := range v.Configuration.EnvVariables {
+			variables[i] = map[string]interface{}{
+				"name":  variable.Name,
+				"value": variable.Value,
+			}
+		}
+
+		result = append(result, map[string]interface{}{
+			"id":       v.EnvironmentId,
+			"variable": variables,
+		})
+	}
+	return result
+}
+
+func flattenComponentIds(list []components.Component) []string {
+	result := make([]string, len(list))
+	for i, component := range list {
+		result[i] = component.ID
+	}
+	return result
+}
+
+func resourceApplicationRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	region := config.GetRegion(d)
+	client, err := config.ServiceStageV2Client(region)
+	if err != nil {
+		return diag.Errorf("error creating ServiceStage v2 client: %s", err)
+	}
+
+	application, err := applications.Get(client, d.Id())
+	if err != nil {
+		common.CheckDeletedDiag(d, err, "error retrieving ServiceStage application")
+	}
+
+	componentList, err := components.List(client, d.Id(), components.ListOpts{})
+	if err != nil {
+		return diag.Errorf("error getting ServiceStage component using application ID (%s): %s", d.Id(), err)
+	}
+
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("name", application.Name),
+		d.Set("description", application.Description),
+		d.Set("enterprise_project_id", application.EnterpriseProjectId),
+		d.Set("component_ids", flattenComponentIds(componentList)),
+	)
+	configs, err := applications.ListConfig(client, d.Id(), applications.ListConfigOpts{})
```

Comment:
what is the differet between component and env config

Reply:
The environment variable logic is written later than the main logic.